File: Emulator/src/main/content/global/skill/magic/spells/modern/GodSpell.java
package content.global.skill.magic.spells.modern;

import core.cache.def.impl.ItemDefinition;
import core.game.container.impl.EquipmentContainer;
import core.game.node.Node;
import core.game.node.entity.Entity;
import core.game.node.entity.combat.BattleState;
import core.game.node.entity.combat.spell.CombatSpell;
import core.game.node.entity.combat.spell.Runes;
import core.game.node.entity.combat.spell.SpellType;
import core.game.node.entity.impl.Animator.Priority;
import core.game.node.entity.impl.Projectile;
import core.game.node.entity.npc.NPC;
import core.game.node.entity.player.Player;
import core.game.node.entity.player.link.SpellBookManager.SpellBook;
import core.game.node.entity.skill.Skills;
import core.game.node.item.Item;
import core.game.world.update.flag.context.Animation;
import core.game.world.update.flag.context.Graphics;
import core.plugin.Initializable;
import core.plugin.Plugin;
import org.rs.consts.Items;
import org.rs.consts.Sounds;

import static core.api.ContentAPIKt.playGlobalAudio;

/**
 * The type God spell.
 */
@Initializable
public final class GodSpell extends CombatSpell {

    private static final String[] NAMES = new String[]{"Saradomin strike", "Guthix claws", "Flames of Zamorak"};
    private static final int[] GOD_STAVES = new int[]{Items.SARADOMIN_STAFF_2415, Items.GUTHIX_STAFF_2416, Items.ZAMORAK_STAFF_2417};
    private static final Graphics SARA_START = null;
    private static final Projectile SARA_PROJECTILE = null;
    private static final Graphics SARA_END = new Graphics(76, 0);
    private static final Graphics GUTHIX_START = null;
    private static final Projectile GUTHIX_PROJECTILE = null;
    private static final Graphics GUTHIX_END = new Graphics(77, 0);
    private static final Graphics ZAM_START = null;
    private static final Projectile ZAM_PROJECTILE = null;
    private static final Graphics ZAM_END = new Graphics(78, 0);
    private static final Animation ANIMATION = new Animation(811, Priority.HIGH);

    /**
     * Instantiates a new God spell.
     */
    public GodSpell() {

    }

    private GodSpell(SpellType type, int sound, int impactAudio, Graphics start, Projectile projectile, Graphics end, Item... runes) {
        super(type, SpellBook.MODERN, 60, 35.0, sound, impactAudio, ANIMATION, start, projectile, end, runes);
    }

    private int getSpellIndex() {
        int index = -1;
        switch (getCastRunes()[1].getAmount()) {
            case 2:
                index = 0;
                break;
            case 1:
                index = 1;
                break;
            case 4:
                index = 2;
                break;
        }
        return index;
    }

    @Override
    public boolean meetsRequirements(Entity caster, boolean message, boolean remove) {
        if (caster instanceof NPC) {
            return true;
        }
        if (caster instanceof Player) {
            int staffId = ((Player) caster).getEquipment().getNew(EquipmentContainer.SLOT_WEAPON).getId();
            int index = getSpellIndex();
            if (index < 0) {
                return false;
            }
            int required = GOD_STAVES[index];
            Player p = (Player) caster;
            if (p.getSavedData().activityData.getGodCasts()[index] < 100 && !p.getZoneMonitor().isInZone("mage arena")) {
                p.sendMessage("You need to cast " + NAMES[index] + " " + (100 - p.getSavedData().activityData.getGodCasts()[index]) + " more times inside the Mage Arena.");
                return false;
            }

            if (staffId != required && !(index == 1 && staffId == Items.VOID_KNIGHT_MACE_8841)) {
                if (message) {
                    ((Player) caster).getPacketDispatch().sendMessage("You need to wear a " + ItemDefinition.forId(required).getName() + "  to cast this spell.");
                }
                return false;
            }
        }
        return super.meetsRequirements(caster, message, remove);
    }

    @Override
    public void fireEffect(Entity entity, Entity victim, BattleState state) {
        switch (getSpellIndex()) {
            case 0:
                victim.getSkills().decrementPrayerPoints(1);
                break;
            case 1:
                victim.getSkills().drainLevel(Skills.DEFENCE, 0.05, 0.05);
                break;
            case 2:
                victim.getSkills().drainLevel(Skills.MAGIC, 0.05, 0.05);
                break;
        }
    }

    @Override
    public void visualize(Entity entity, Node target) {
        super.visualize(entity, target);
        if (entity instanceof NPC) {
            NPC n = (NPC) entity;
            if (n.getId() > 911 && n.getId() < 915 || (n.getId() > 906 && n.getId() < 912)) {
                n.getAnimator().forceAnimation(n.getProperties().getAttackAnimation());
            }
        }
    }

    @Override
    public void visualizeImpact(Entity entity, Entity target, BattleState state) {
        if (entity instanceof Player) {
            int index = getSpellIndex();
            Player p = (Player) entity;
            if (p.getSavedData().activityData.getGodCasts()[index] < 100) {
                p.getSavedData().activityData.getGodCasts()[index]++;
                if (p.getSavedData().activityData.getGodCasts()[index] >= 100) {
                    p.sendMessage("You can now cast " + NAMES[index] + " outside the Arena.");
                }
            }
            if (state.getEstimatedHit() == -1) {
                target.graphics(SPLASH_GRAPHICS);
                if (projectile == SARA_PROJECTILE) {
                    playGlobalAudio(target.getLocation(), Sounds.SARADOMIN_STRIKE_FAIL_1656, 20);
                }
                if (projectile == GUTHIX_PROJECTILE) {
                    playGlobalAudio(target.getLocation(), Sounds.CLAWS_OF_GUTHIX_FAIL_1652, 20);
                }
                if (projectile == ZAM_PROJECTILE) {
                    playGlobalAudio(target.getLocation(), Sounds.FLAMES_OF_ZAMORAK_FAIL_1654, 20);
                }
                return;
            }
        }
        target.graphics(endGraphics);
        playGlobalAudio(target.getLocation(), impactAudio);
    }

    @Override
    public int getMaximumImpact(Entity entity, Entity victim, BattleState state) {
        return getType().getImpactAmount(entity, victim, 0);
    }

    @Override
    public Plugin<SpellType> newInstance(SpellType type) throws Throwable {
        SpellBook.MODERN.register(41, new GodSpell(SpellType.GOD_STRIKE, -1, Sounds.SARADOMIN_STRIKE_1659, SARA_START, SARA_PROJECTILE, SARA_END, Runes.BLOOD_RUNE.getItem(2), Runes.FIRE_RUNE.getItem(2), Runes.AIR_RUNE.getItem(4)));
        SpellBook.MODERN.register(42, new GodSpell(SpellType.GOD_STRIKE, -1, Sounds.CLAWS_OF_GUTHIX_1653, GUTHIX_START, GUTHIX_PROJECTILE, GUTHIX_END, Runes.BLOOD_RUNE.getItem(2), Runes.FIRE_RUNE.getItem(1), Runes.AIR_RUNE.getItem(4)));
        SpellBook.MODERN.register(43, new GodSpell(SpellType.GOD_STRIKE, -1, Sounds.FLAMES_OF_ZAMORAK_1655, ZAM_START, ZAM_PROJECTILE, ZAM_END, Runes.BLOOD_RUNE.getItem(2), Runes.FIRE_RUNE.getItem(4), Runes.AIR_RUNE.getItem(1)));
        return this;
    }

}

File: Emulator/src/main/content/global/skill/magic/spells/modern/GodSpell.kt
package content.global.skill.magic.spells.modern

import content.global.skill.magic.spells.ModernSpells
import core.api.playGlobalAudio
import core.api.sendMessage
import core.cache.def.impl.ItemDefinition
import core.game.container.impl.EquipmentContainer
import core.game.node.Node
import core.game.node.entity.Entity
import core.game.node.entity.combat.BattleState
import core.game.node.entity.combat.spell.CombatSpell
import core.game.node.entity.combat.spell.SpellType
import core.game.node.entity.impl.Animator.Priority
import core.game.node.entity.npc.NPC
import core.game.node.entity.player.Player
import core.game.node.entity.player.link.SpellBookManager.SpellBook
import core.game.world.update.flag.context.Animation
import core.game.world.update.flag.context.Graphics
import core.plugin.Initializable
import core.plugin.Plugin
import shared.consts.Animations
import shared.consts.Items
import shared.consts.NPCs

/**
 * Represents the God spell.
 */
@Initializable
class GodSpell private constructor(
    private val data: GodSpellDefinition
) : CombatSpell(
    SpellType.GOD_STRIKE,
    SpellBook.MODERN,
    60,
    35.0,
    -1,
    data.impactAudio,
    ANIMATION,
    null,
    data.projectile,
    data.endGraphics,
    *data.runes
) {
    constructor() : this(GodSpellDefinition.SARADOMIN)

    override fun fireEffect(entity: Entity, victim: Entity, state: BattleState) {
        data.effect(entity, victim)
    }

    override fun visualize(entity: Entity, target: Node) {
        super.visualize(entity, target)
        if (entity is NPC) {
            val id = entity.id
            if ((id in (NPCs.KOLODION_911 + 1) until NPCs.LEELA_915) || (id in (NPCs.KOLODION_906 + 1) until NPCs.BATTLE_MAGE_912)) {
                entity.animator.forceAnimation(entity.properties.attackAnimation)
            }
        }
    }

    override fun visualizeImpact(entity: Entity, target: Entity, state: BattleState) {
        if (entity is Player) {
            val p = entity
            val index = INDEX_CACHE[data] ?: run {
                val idx = GOD_SPELLS.indexOf(data)
                INDEX_CACHE[data] = idx
                idx
            }
            val castData = p.getSavedData().activityData
            val spellName = getSpellName(data.name)
            if (castData.godCasts[index] < data.requiredCasts) {
                castData.godCasts[index]++
                if (castData.godCasts[index] == data.requiredCasts) {
                    sendMessage(p, "You can now cast $spellName outside the Arena.")
                }
            }

            if (state.estimatedHit == -1) {
                target.graphics(SPLASH_GRAPHIC)
                playGlobalAudio(target.location, data.failAudio, 20)
                return
            }
        }

        target.graphics(data.endGraphics)
        playGlobalAudio(target.location, data.impactAudio)
    }

    override fun meetsRequirements(caster: Entity, message: Boolean, remove: Boolean): Boolean {
        if (caster !is Player) return true

        val staffId = caster.equipment.getNew(EquipmentContainer.SLOT_WEAPON).id
        val castData = caster.getSavedData().activityData
        val index = INDEX_CACHE[data] ?: run {
            val idx = GOD_SPELLS.indexOf(data)
            INDEX_CACHE[data] = idx
            idx
        }
        val spellName = getSpellName(data.name)
        if (castData.godCasts[index] < data.requiredCasts && !caster.zoneMonitor.isInZone("mage arena")) {
            sendMessage(caster, "You need to cast $spellName ${data.requiredCasts - castData.godCasts[index]} more times inside the Mage Arena.")
            return false
        }

        val hasCorrectStaff =
            staffId == data.staffId || (data == GodSpellDefinition.GUTHIX && staffId == Items.VOID_KNIGHT_MACE_8841)
        if (!hasCorrectStaff) {
            if (message) {
                sendMessage(caster, "You need to wear a ${ItemDefinition.forId(data.staffId).name} to cast this spell.")
            }
            return false
        }

        return super.meetsRequirements(caster, message, remove)
    }

    override fun getMaximumImpact(entity: Entity, victim: Entity, state: BattleState): Int {
        return getType().getImpactAmount(entity, victim, 0)
    }

    private fun getSpellName(name: String): String {
        return when (name) {
            "SARADOMIN" -> "Saradomin strike"
            "ZAMORAK" -> "Flames of Zamorak"
            else -> "Claws of Guthix"
        }
    }

    override fun newInstance(type: SpellType?): Plugin<SpellType?> {
        val buttons = intArrayOf(
            ModernSpells.SARADOMIN_STRIKE,
            ModernSpells.CLAWS_OF_GUTHIX,
            ModernSpells.FLAMES_OF_ZAMORAK,
        )
        for (i in buttons.indices) {
            SpellBook.MODERN.register(buttons[i], GodSpell(GOD_SPELLS[i]))
        }
        return this
    }

    companion object {
        private val ANIMATION = Animation(Animations.HUMAN_CAST_SPELL_LONG_811, Priority.HIGH)
        private val SPLASH_GRAPHIC = Graphics(85, 0, 100)
        private val GOD_SPELLS = enumValues<GodSpellDefinition>()
        private val INDEX_CACHE = mutableMapOf<GodSpellDefinition, Int>()
    }
}
